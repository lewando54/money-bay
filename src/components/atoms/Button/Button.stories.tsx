/* eslint-disable @typescript-eslint/consistent-type-assertions */
import { Meta,StoryObj } from '@storybook/react-native'

import Button from './Button'

import { globalColors } from '@styling/GlobalTheme.style'

const meta = {
    component: Button,
    argTypes: {
        color: { control: 'select', options: ['primary', 'secondary'] },
        size: { control: 'select', options: ['normal', 'small'] }
    }
} satisfies Meta<typeof Button>

export default meta
type Story = StoryObj<typeof Button>

export const Primary: Story = {
    args: {
        color: 'primary',
        size: 'normal',
        children: 'Button'
    }
}

export const Secondary: Story = {
    args: {
        color: 'secondary',
        size: 'normal',
        children: 'Button'
    }
}

export const Small: Story = {
    args: {
        color: globalColors.GOOD_GREEN,
        size: 'small',
        children: 'Button'
    }
}