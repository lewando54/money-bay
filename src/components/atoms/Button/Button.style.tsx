import { StyleSheet } from 'react-native'
export default StyleSheet.create({
    button: {
        borderRadius: 10,
        padding: 15
    },
    primary: {

    },
    secondaryButton: {
        borderColor: '#CED6E1', 
        borderWidth: 1
    },
    text: {
        fontSize: 16,
        textAlign: 'center',
        fontFamily: 'MulishSemiBold'
    },
    primaryText: {
        color: 'white'
    },
    secondaryText: {
        color: '#1B1D4D'
    }
})