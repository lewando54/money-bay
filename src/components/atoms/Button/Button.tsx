import React from 'react'
import { Text, TouchableOpacity, GestureResponderEvent, ViewStyle, StyleProp } from 'react-native'
import { LinearGradient } from 'expo-linear-gradient'
import buttonStyle from './Button.style'
import GlobalThemeStyle, { PRIMARY_GRADIENT_END_COLOR, PRIMARY_GRADIENT_START_COLOR, SECONDARY_GRADIENT_FULL_COLOR } from '../../../styling/GlobalTheme.style'

type colorProp = 'primary' | 'secondary' | string
type sizeProp = 'normal' | 'small'

interface IButtonProps{
    children: React.ReactNode
    color?: colorProp
    size?: sizeProp
    width?: string
    containerStyle?: StyleProp<ViewStyle>
    onClick?: ((event: GestureResponderEvent) => void) | null | undefined
    testID?: string
}

export default function Button({
    children,
    color = 'primary',
    size = 'normal',
    width = '100%',
    containerStyle,
    onClick,
    testID
}: IButtonProps){
    const gradientColor = 
        color === 'primary' ? [PRIMARY_GRADIENT_START_COLOR, PRIMARY_GRADIENT_END_COLOR]
            : color === 'secondary' ? [SECONDARY_GRADIENT_FULL_COLOR, SECONDARY_GRADIENT_FULL_COLOR]
                : [color, color]

    return (
        <TouchableOpacity style={[{width: width}, containerStyle]} onPress={onClick} testID={testID}>
            <LinearGradient 
                colors={gradientColor}
                start={{x: 0, y: 0.5}}
                end={{x: 1, y: 0.5}}
                style={[
                    buttonStyle.button,
                    size === 'small' && { padding: 8, borderRadius: 5 },
                    color === 'secondary' && buttonStyle.secondaryButton
                ]}>
                <Text style={[
                    buttonStyle.text,
                    color === 'primary' && buttonStyle.primaryText,
                    color === 'secondary' && buttonStyle.secondaryText,
                    size === 'small' && [{ fontSize: 12 }, GlobalThemeStyle.text_SemiBold]
                ]}>
                    {children}
                </Text>
            </LinearGradient>
        </TouchableOpacity>
    )
}