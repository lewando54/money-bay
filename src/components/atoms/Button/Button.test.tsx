import React from 'react'
import renderer from 'react-test-renderer'
import { render, fireEvent, screen } from '@testing-library/react-native'
import Button from './Button'
import { PRIMARY_GRADIENT_END_COLOR, PRIMARY_GRADIENT_START_COLOR, SECONDARY_GRADIENT_FULL_COLOR } from '../../../styling/GlobalTheme.style'

describe('Button', () => {
  it('should render correctly', () => {
    const tree = renderer.create(<Button testID='testxd' color="primary">Click me</Button>).toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('should render with primary color by default', () => {
    const tree = renderer.create(<Button testID='testxd'>Click me</Button>).toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('should render innerHTML text', () => {
    const { getByText } = render(<Button testID='testxd' color="secondary">Click me</Button>)
    expect(getByText('Click me')).toBeTruthy()
  })

  it('should render with primary gradient colors', () => {
    const { getByTestId } = render(<Button testID='button' color={'primary'}>Click me</Button>)
    expect( getByTestId('button').props.children[0].props.colors[0]).toEqual(PRIMARY_GRADIENT_START_COLOR)
    expect( getByTestId('button').props.children[0].props.colors[1]).toEqual(PRIMARY_GRADIENT_END_COLOR)
  })

  it('should render with secondary color', () => {
    const { getByTestId } = render(<Button testID='button' color={'secondary'}>Click me</Button>)
    expect( getByTestId('button').props.children[0].props.colors[0]).toEqual(SECONDARY_GRADIENT_FULL_COLOR)
    expect( getByTestId('button').props.children[0].props.colors[1]).toEqual(SECONDARY_GRADIENT_FULL_COLOR)
  })

  it('should render with custom color', () => {
    render(<Button testID='button' color={'#fff'}>Click me</Button>)
    expect(screen.getByTestId('button').props.children[0].props.colors[0]).toEqual('#fff')
    expect(screen.getByTestId('button').props.children[0].props.colors[1]).toEqual('#fff')
  })

  it('should render small variant', () => {
    const tree = renderer.create(<Button testID='testxd' color="primary" size={'small'}>Click me</Button>).toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('should render normal variant', () => {
    const tree = renderer.create(<Button testID='testxd' color="primary" size={'normal'}>Click me</Button>).toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('should call onClick function when clicked', () => {
    const onClick = jest.fn()
    const { getByTestId } = render(<Button testID='testxd' onClick={onClick}>Click me</Button>)
    fireEvent.press(getByTestId('testxd'))
    expect(onClick).toHaveBeenCalled()
  })
})
