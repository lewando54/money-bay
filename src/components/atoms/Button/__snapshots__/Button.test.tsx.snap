// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button should render correctly 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": undefined,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={false}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "opacity": 1,
      "width": "100%",
    }
  }
  testID="test"
>
  <ViewManagerAdapter_ExpoLinearGradient
    proxiedProperties={
      {
        "colors": [
          4288142079,
          4279654301,
        ],
        "endPoint": [
          1,
          0.5,
        ],
        "locations": undefined,
        "startPoint": [
          0,
          0.5,
        ],
      }
    }
    style={
      [
        {
          "borderRadius": 10,
          "padding": 15,
        },
        false,
        false,
      ]
    }
  >
    <Text
      style={
        [
          {
            "color": "#1B1D4D",
            "fontFamily": "MulishRegular",
            "fontSize": 14,
          },
          [
            {
              "fontFamily": "MulishSemiBold",
              "fontSize": 16,
              "textAlign": "center",
            },
            {
              "color": "white",
            },
            false,
            false,
          ],
        ]
      }
    >
      Click me
    </Text>
  </ViewManagerAdapter_ExpoLinearGradient>
</View>
`;

exports[`Button should render normal variant 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": undefined,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={false}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "opacity": 1,
      "width": "100%",
    }
  }
  testID="test"
>
  <ViewManagerAdapter_ExpoLinearGradient
    proxiedProperties={
      {
        "colors": [
          4288142079,
          4279654301,
        ],
        "endPoint": [
          1,
          0.5,
        ],
        "locations": undefined,
        "startPoint": [
          0,
          0.5,
        ],
      }
    }
    style={
      [
        {
          "borderRadius": 10,
          "padding": 15,
        },
        false,
        false,
      ]
    }
  >
    <Text
      style={
        [
          {
            "color": "#1B1D4D",
            "fontFamily": "MulishRegular",
            "fontSize": 14,
          },
          [
            {
              "fontFamily": "MulishSemiBold",
              "fontSize": 16,
              "textAlign": "center",
            },
            {
              "color": "white",
            },
            false,
            false,
          ],
        ]
      }
    >
      Click me
    </Text>
  </ViewManagerAdapter_ExpoLinearGradient>
</View>
`;

exports[`Button should render small variant 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": undefined,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={false}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "opacity": 1,
      "width": "100%",
    }
  }
  testID="test"
>
  <ViewManagerAdapter_ExpoLinearGradient
    proxiedProperties={
      {
        "colors": [
          4288142079,
          4279654301,
        ],
        "endPoint": [
          1,
          0.5,
        ],
        "locations": undefined,
        "startPoint": [
          0,
          0.5,
        ],
      }
    }
    style={
      [
        {
          "borderRadius": 10,
          "padding": 15,
        },
        {
          "borderRadius": 5,
          "padding": 8,
        },
        false,
      ]
    }
  >
    <Text
      style={
        [
          {
            "color": "#1B1D4D",
            "fontFamily": "MulishRegular",
            "fontSize": 14,
          },
          [
            {
              "fontFamily": "MulishSemiBold",
              "fontSize": 16,
              "textAlign": "center",
            },
            {
              "color": "white",
            },
            false,
            [
              {
                "fontSize": 12,
              },
              {
                "color": "white",
                "fontFamily": "MulishSemiBold",
              },
            ],
          ],
        ]
      }
    >
      Click me
    </Text>
  </ViewManagerAdapter_ExpoLinearGradient>
</View>
`;

exports[`Button should render with primary color by default 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": undefined,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={false}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "opacity": 1,
      "width": "100%",
    }
  }
  testID="test"
>
  <ViewManagerAdapter_ExpoLinearGradient
    proxiedProperties={
      {
        "colors": [
          4288142079,
          4279654301,
        ],
        "endPoint": [
          1,
          0.5,
        ],
        "locations": undefined,
        "startPoint": [
          0,
          0.5,
        ],
      }
    }
    style={
      [
        {
          "borderRadius": 10,
          "padding": 15,
        },
        false,
        false,
      ]
    }
  >
    <Text
      style={
        [
          {
            "color": "#1B1D4D",
            "fontFamily": "MulishRegular",
            "fontSize": 14,
          },
          [
            {
              "fontFamily": "MulishSemiBold",
              "fontSize": 16,
              "textAlign": "center",
            },
            {
              "color": "white",
            },
            false,
            false,
          ],
        ]
      }
    >
      Click me
    </Text>
  </ViewManagerAdapter_ExpoLinearGradient>
</View>
`;
